### PR TITLE
Liquidity Rewards contract

### PR DESCRIPTION
**Pull request by crazybuster**

### CODE REVIEW

**Review by sparrowDom — APPROVED** [review]
I am already integrating this so from API standpoint things look fine for now. Will defer to others when it comes to more in-depth review of solidity code.

File: contracts/contracts/liquidity/LiquidityReward.sol [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+    event CampaignStarted(
+        uint256 rewardRate,
+        uint256 startBlock,
+        uint256 endBlock
+    );
+    event CampaignStopped(uint256 endBlock);
+    event Deposit(address indexed user, uint256 amount);
+    event Withdraw(address indexed user, uint256 amount);
+    event Claim(address indexed user, uint256 amount);
+    event EmergencyWithdraw(address indexed user, uint256 amount);
```
Comment (franckc) [inline contracts/contracts/liquidity/LiquidityReward.sol:74]: I don't see the event EmergencyWithdraw used anywhere?

**Review by franckc — COMMENTED** [review]
Took a quick pass at reviewing. Looks great. A few minor comments inline.
I'll look more in-depth.

File: contracts/contracts/liquidity/LiquidityReward.sol [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+    function stopCampaign() external onlyGovernor {
+      //calculate until current pool
+      updatePool();
+      //end the block here (the CampaignMultiplier will be zero)
+      endBlock = block.number;
+      emit CampaignStopped(endBlock);
+    }
+
+    function campaignActive() external view returns (bool) {
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/liquidity/LiquidityReward.sol:146]: I still think we need this change.

File: contracts/contracts/liquidity/LiquidityReward.sol (hunk 1 above)
Comment (DanielVF) [reply contracts/contracts/liquidity/LiquidityReward.sol:74]: Yes, lets get rid of this event.

**Review by DanielVF — APPROVED** [review]

File: contracts/contracts/liquidity/LiquidityReward.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (crazybuster) [reply contracts/contracts/liquidity/LiquidityReward.sol:145]: This isn't actually used anywhere, but will update it to be correct

File: contracts/contracts/liquidity/LiquidityReward.sol [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    function emergencyWithdraw() public {
+        UserInfo storage user = userInfo[msg.sender];
+        uint256 amount = user.amount;
+        user.amount = 0;
+        user.rewardDebt = 0;
+        pool.lpToken.safeTransfer(address(msg.sender), amount);
+        emit EmergencyWithdraw(msg.sender, amount);
+    }
+
+    function subDebt(uint256 amount, int256 debt) internal pure returns (uint256 result) {
```
[the remark this answers is not in the document]
Comment (crazybuster) [reply contracts/contracts/liquidity/LiquidityReward.sol]: I believe subDebt is using safeMath in terms of sub and add

File: contracts/contracts/liquidity/LiquidityReward.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (crazybuster) [reply contracts/contracts/liquidity/LiquidityReward.sol:145]: This does check if we're past end block or not, what it didn't do was check if we're at or past the lastRewardBlock which will be set to the future if the startBlock is at a later date.

**Pull request merged**